File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
pair<int, int> dfs(TreeNode* node) {
    if (!node) {
        return {0, 0};
    }
    pair<int, int> left = dfs(node->left);
    pair<int, int> right = dfs(node->right);

    int diamWithNode = 1 + left.second + right.second;
    int maxBranch = 1 + max(left.second, right.second);

    return {max(max(left.first, right.first), diamWithNode), maxBranch};
}
public:
    int diameterOfBinaryTree(TreeNode* root) {
        return dfs(root).first - 1;
    }
};

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        small, large = p.val, q.val
        if small > large:
            small, large = large, small
        
        if small <= root.val and root.val <= large:
            return root
        if root.val < small:
            return self.lowestCommonAncestor(root.right, p, q)
        return self.lowestCommonAncestor(root.left, p, q)